fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/context/ContainerItemContext.java

```diff
@@ -0,0 +1,161 @@
+ * <ul>
+ *     <li>Query how much of the variant is in the inventory through {@link #getAmount}.</li>
+ *     <li>Transform some of the bound variant into another variant through {@link #transform}.</li>
+ * </ul>
+ */
+// TODO: update javadoc
+// TODO: wrapper around a SingleSlotStorage<ItemVariant>
+// TODO: wrapper around a StackReference maybe?
+public interface ContainerItemContext {
+	/**
+	 * Try to find an API instance for the passed lookup and return it, or {@code null} if there is none.
+	 * The API is queried for the current variant, if it's not blank.
+	 *
+	 * @see ItemApiLookup#find
+	 */
+	@Nullable
+	default <A> A find(ItemApiLookup<A, ContainerItemContext> lookup) {
+		return getItemVariant().isBlank() ? null : lookup.find(getItemVariant().toStack(), this);
+	}
+
+	/**
+	 * Return the current item variant of this context, that is the variant in the slot of the context.
+	 * If the result is non blank, {@link #getAmount} should be
+	 */
+	ItemVariant getItemVariant();
+
+	/**
+	 * Return the current amount of {@link #getItemVariant()} in the slot of the context.
+	 *
+	 * @throws IllegalStateException If {@linkplain #getItemVariant() the current variant} is blank.
+	 */
+	long getAmount();
+
+	/**
+	 * Try to insert some items into this context, prioritizing the slot of the context over the rest of the inventory.
+	 *
+	 * @see Storage#insert
+	 */
+	long insert(ItemVariant itemVariant, long maxAmount, TransactionContext transaction);
+
+	/**
+	 * Try to extract some items from this context's slot.
+	 *
+	 * @see Storage#extract
+	 */
+	long extract(ItemVariant itemVariant, long maxAmount, TransactionContext transaction);
+
+	/**
+	 * Try to transform as many as possibly of {@linkplain #getItemVariant() the current variant} into another variant.
+	 * That is, extract the old variant, and insert the same amount of the new variant instead.
+	 *
+	 * @param into The variant of the items after the conversion. May not be blank.
+	 * @param maxAmount The maximum amount of items to convert. May not be negative.
+	 * @param transaction The transaction this operation is part of.
+	 * @return A nonnegative integer not greater than maxAmount: the amount that was transformed.
+	 */
+	default long transform(ItemVariant into, long maxAmount, TransactionContext transaction) {
+		StoragePreconditions.notBlankNotNegative(into, maxAmount);
+
+		try (Transaction nested = transaction.openNested()) {
+			long extracted = extract(getItemVariant(), maxAmount, nested);
+
+			if (insert(into, maxAmount, nested) == extracted) {
+				nested.commit();
+				return extracted;
+			}
+		}
+
+		return 0;
+	}
+
+	/**
+	 * Get additional slots that may be available in this context.
+	 * This may or may not include the slot of this context.
+	 *
+	 * @return An unmodifiable list containing additional slots of this context.
+	 */
+	List<SingleSlotStorage<ItemVariant>> getInventorySlots();
+
+	/**
+	 * Return the world of this context.
+	 */
+	World getWorld();
```

Comment:
imo this should be left out, if the position isn't provided, the world shouldn't be either. And the world/position isn't always known either. I would instead have an API API for this or something else

Reply:
Position is terrible, agreed. It just feels wrong.

However `World` makes sense to access persistent state and similar global state I think. Otherwise ender buckets have to rely on lifecycle events to capture server instances, and get the client world through the `MinecraftClient` singleton. Not yet totally sure, quite low cost for players though (the entity contains the world already).

Reply:
Removed `getWorld()`.